eeps/eep-0058.md

```diff
@@ -0,0 +1,255 @@
+    Author: Sergey Prokhorov <seriy(dot)pr(at)gmail(dot)com>
+    Status: Draft
+    Type: Standards Track
+    Created: 14-Sep-2021
+    Erlang-Version: 25
+    Post-History:
+****
+EEP 58: Map comprehensions
+----
+
+
+Abstract
+========
+
+This EEP proposes syntax and semantics for map comprehensions and
+generators, similar to list and binary comprehensions and generators,
+but operating on maps and map iterators.
+
+Copyright
+=========
+
+This document has been placed in the public domain.
+
+
+Specification
+=============
+
+Maps and the initial map comprehension syntax and semantics were proposed
+in EEP-43. List and binary comprehensions were specified post-fact
+in EEP-12. This EEP is based on those two EEPs.
+
+The syntax construct commonly referred as a "comprehension" is, in fact,
+composed of 3 semi-independent parts:
+
+- the "generator" - is the expression that iterates over the container,
+  matches each element of the container over a pattern and binds the
+  variables which can be used in the "comprehensions" and the "filters".
+- the "comprehension" - is the expression that is executed for each
+  element produced by the generator that successfully matched the pattern
+  and passed the "filter" check. The value produced by this expression
+  is appended to the resulting generated container.
+- the "filter" - is the expression that is evaluated for each element
+  produced by the generator that matched the pattern and returns a
+  boolean result: when it returns `false`, the current element is
+  skipped from being passed to the "comprehension"; when it results
+  in `true`, the element is passed to the "comprehension".
+
+So, in this document we are going to keep this separation, but
+sometimes will refer to the whole combination of the comprehension, the filter and
+the generator as a "comprehension", because the generator syntax is not
+used outside of the comprehension context.
+
+
+### Syntax
+
+The proposed syntax for the map comprehension part is
+
+    '#{' KeyExpr '=>' ValueExpr '||' GeneratorsFilters '}'
```

Comment:
There is, well, let's call it an inconvenience, just pointing it out: you can't generate `KeyExpr => ValueExpr` "in one".

For example, while something like this is possible with a list comprehension:
```erlang
%% (double all even numbers and tag as even or odd)
[ if N rem 2 =:= 0 -> {N*2, even}; true -> {N, odd} end || N <- lists:seq(1, 10) ]
```

with map comprehensions, this isn't:
```erlang
#{ if N rem 2 =:= 0 -> N*2 => even; true -> N => odd end || N <- lists:seq(1, 10) }
```
but has to be done with two separate `if`s:
```erlang
#{ if N rem 2 =:= 0 -> N*2; true -> N end => if N rem 2 =:= 0 -> even; true -> odd end; || N <- lists:seq(1, 10) }
```

Reply:
yes, and I've been thinking a lot about this restriction and didn't manage to come with any real solution. Maybe some crazy stuff like defining the comprehension as
```
'#{' <expr> ',' <key expr> => <value expr> '||' ...}
```
so you can write smth like

```
#{
  if N rem 2 =:= 0 -> Key = N*2, Val = even;
     true -> Key = N, Val = odd
  end,
  Key => Val
  || N <- lists:seq(1, 10)
}
```

Reply:
This already is valid:
```erlang
[ {K,V} || N <- lists:seq(1, 10),
       {K,V} <- [if N rem 2 =:= 0 -> {N*2,even}; true -> {N,odd} end]].
```
So this should be as well:
```erlang
#{ K => V || N <- lists:seq(1, 10),
         {K,V} <- [if N rem 2 =:= 0 -> {N*2,even}; true -> {N,odd} end]}
```

Reply:
@fenollp yep, this would work with the spec I'm proposing. But it's really not obvious construction IMO

Reply:
Whoa... I think that looks seriously messy 🤪 I wouldn't really want that, TBH.

Looking at binary comprehensions, the comprehension part must (?) result in a binary/bitstring, for example:
```erlang
> << <<(X+1):8>> || <<X:8>> <= <<"abc">> >>.
<<"bcd">>
```

What if, in a similar manner, the map comprehensions' comprehension part must result in a map instead of a `K => V` construct, which is used to update the already generated map? Aside from solving what I outlined above, this would even allow to insert _multiple_ key-value pairs at a point, as well _not_ inserting anything at all. Again looking at binary comprehensions, both
```erlang
> << <<(X+1):8, (X+2):8)>> || <<X:8>> <= <<"abc">> >>.
<<"bccdde">>
```
and
```erlang
> << <<>> || <<X:8>> <= <<"abc">> >>.
<<>>
```
are allowed.

Reply:
Good point. I don't think it's a problem though. There are always things you can't do and ways to work around it.

Here's a workaround using a "singleton-generator":

```Erlang
[ K => V || N <- lists:seq(1, 10),
            {K, V} <- [if N rem 2 =:= 0 -> {N*2, even}; true -> {N, odd} end] ]
```

Reply:
@fennolp @zuiderkwast yes, but as @seriyps pointed out, that's not very obvious. I would call it awkward-hackish even. Aside from that, it requires that secondary helper generator. The compiler might optimize things here, but I doubt it TBH.

I'm not too empathic about this issue actually, I just like to point things out _before_ a ship has sailed ;)

Reply:
There are other things you can't do, such as emitting two elements each round, and workaround using extra generator.

* Example: `[2*N, 2*N + 1 || N <- lists:seq(1, 10)]`
* Workaround: `[X || N <- lists:seq(1, 10), X <- [2*N, 2*N + 1]]`

Reply:
My point, actually: that ship has sailed. And, at a wild guess, not because it was rejected, but because nobody considered it back then. It is possible to do, with a fold or explicit recursion if nothing else, but if we can do it nicely and there is no reason to reject it, why not _do_ it nicely? 😉

Reply:
Oh, I didn't know it's possible with binary comprehensions:
```erlang
> << case (C rem 2) == 0 of true -> <<1>>; false -> <<2>> end || C <- lists:seq(1, 10) >>.
<<2,1,2,1,2,1,2,1,2,1>>
> << case (C rem 2) == 0 of true -> <<1>>; false -> oups end || C <- lists:seq(1, 10) >>. 
** exception error: bad argument
     in function  bit_size/1
        called as bit_size(oups)
```

In that case maybe it makes sense to offer 2 versions of map comprehension: short one
```erlang
'#{' KeyExpr '=>' ValueExpr '||' GeneratorsFilters '}'
```
but then also the one that yields arbitrary map
```erlang
'#{' AnyExprThatReturnsMap '||' GeneratorsFilters '}'
```
it's going to be trickier to optimize though. With short version it's possible to just allocate KeyExpr and ValueExpr to BEAM registers leaving no garbage on heap, but with the 2nd version I guess it would be difficult to avoid creation of temporary maps. I would also be disappointed if programmers start to write
```erlang
#{ #{K => V * 2} || K := V <- Map}
```
just because it's possible or they don't remember what way is correct =)

Reply:
> In that case maybe it makes sense to offer 2 versions of map comprehension

I'm not a huge fan of "there is more than one way to do it", but it might make sense here.

That said, I have only now glimpsed at the code in the reference PR, and while I can't say I grasp it fully, it looks like you turn the map comprehension into a list comprehension, and the resulting `{K, V}` list into a map via `maps:from_list/1`, yes?

In that case, allowing maps as returns would indeed make things difficult. As I see it, instead of a simple `maps:from_list/1` (which is a BIF, btw), you would have to resort to `lists:foldl/3`, like
```erlang
lists:foldl(
    fun
        ({Key, Value}, Acc) -> Acc#{Key => Value};
        (Map, Acc) -> maps:merge(Acc, Map)
    end,
    #{},
    ListCompResult
)
```

Reply:
Our PR was done as a one-day hackathon, so, we have choosen the easiest way it could be done. But eventually it would be much better if we implement a distinct core-erlang code for map comprehensions. Especially given that `{K, V} <- maps:to_list(Map)` would not work for `maps:iterator()`. Also, I'm sure it can be done more efficiently (producing less garbage on heap) if we avoid construction of intermediate tuples and lists in the implementation; but this may be done in a follow-up I think.